File: android/app/src/main/kotlin/com/sps/patrol/gps/SuperGpsLocationEngine.kt
package com.sps.patrol.gps

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import com.google.android.gms.location.CurrentLocationRequest
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.google.android.gms.tasks.CancellationTokenSource
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

internal object SuperGpsLocationEngine {
    /** Chấp nhận cache lastLocation nếu mới hơn 2 phút — trả về tức thì. */
    private const val FAST_CACHE_MAX_AGE_MS = 120_000L

    /** getCurrentLocation: dùng cache tới 60s, tối đa chờ GPS mới 1.5s. */
    private const val CURRENT_LOC_MAX_AGE_MS = 60_000L
    private const val CURRENT_LOC_MAX_WAIT_MS = 1_500L

    /** One-shot: cache nhanh chỉ khi accuracy đủ tốt (sau Kalman). */
    private const val ONE_SHOT_BEST_ACCURACY_M = 4f

    private val kalman = SuperGpsKalmanFilter()
    private val fixGate = SuperGpsFixQualityGate()
    private val barometer = SuperGpsBarometer()
    private val sinks = linkedSetOf<EventChannel.EventSink>()

    private var fusedClient: FusedLocationProviderClient? = null
    private var locationCallback: LocationCallback? = null
    private var running = false
    private var streamOptions = SuperGpsStreamOptions()
    private var startGeneration = 0

    @Synchronized
    fun addListener(
        context: Context,
        sink: EventChannel.EventSink,
        options: SuperGpsStreamOptions = SuperGpsStreamOptions(),
    ) {
        val optionsChanged = streamOptions != options
        streamOptions = options
        sinks.add(sink)
        when {
            !running -> start(context.applicationContext)
            optionsChanged -> restartUpdates(context.applicationContext)
        }
    }

    @Synchronized
    fun removeListener(sink: EventChannel.EventSink) {
        sinks.remove(sink)
        if (sinks.isEmpty()) {
            stop()
        }
    }

    fun isBarometerHardwareSupported(context: Context): Boolean {
        barometer.bind(context.applicationContext)
        return barometer.hasHardwareSupport
    }

    fun getCurrentPosition(
        context: Context,
        enableBarometer: Boolean,
        result: MethodChannel.Result,
    ) {
        if (!hasFineLocation(context)) {
            result.error("PERMISSION_DENIED", "Location permission not granted", null)
            return
        }

        val appContext = context.applicationContext
        barometer.bind(appContext)
        val shouldRunBaro = enableBarometer && barometer.hasHardwareSupport
        val client = LocationServices.getFusedLocationProviderClient(context)

        if (shouldRunBaro) {
            if (!barometer.isActive) {
                barometer.start(appContext)
            }
            Thread {
                barometer.awaitFirstReading(SuperGpsBarometer.FIRST_READING_TIMEOUT_MS)
                Handler(Looper.getMainLooper()).post {
                    fetchCurrentPositionGps(client, shouldRunBaro, result)
                }
            }.start()
            return
        }

        fetchCurrentPositionGps(client, shouldRunBaro = false, result)
    }

    private fun fetchCurrentPositionGps(
        client: FusedLocationProviderClient,
        shouldRunBaro: Boolean,
        result: MethodChannel.Result,
    ) {
        synchronized(this) {
            if (!running) {
                kalman.reset()
                fixGate.reset()
            }
        }

        // 1) lastLocation — thường <50ms nếu app/OS đã có cache.
        client.lastLocation
            .addOnSuccessListener { cached ->
                if (cached != null &&
                    isLocationFresh(cached, FAST_CACHE_MAX_AGE_MS) &&
                    cached.accuracy <= ONE_SHOT_BEST_ACCURACY_M
                ) {
                    val payload = buildOneShotPayloadIfAccepted(
                        cached,
                        includeBarometer = shouldRunBaro,
                    )
                    if (payload != null) {
                        finishGetCurrentPosition(
                            shouldRunBaro,
                            result,
                            payload,
                        )
                        return@addOnSuccessListener
                    }
                }
                requestCurrentLocationBounded(client, cached, shouldRunBaro, result)
            }
            .addOnFailureListener {
                requestCurrentLocationBounded(client, null, shouldRunBaro, result)
            }
    }

    private fun finishGetCurrentPosition(
        stopBarometerAfter: Boolean,
        result: MethodChannel.Result,
        payload: Map<String, Any?>?,
    ) {
        if (stopBarometerAfter && !running) {
            barometer.stop()
        }
        result.success(payload)
    }

    private fun requestCurrentLocationBounded(
        client: FusedLocationProviderClient,
        staleFallback: Location?,
        stopBarometerAfter: Boolean,
        result: MethodChannel.Result,
    ) {
        val request = CurrentLocationRequest.Builder()
            .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
            .setMaxUpdateAgeMillis(CURRENT_LOC_MAX_AGE_MS)
            .setDurationMillis(CURRENT_LOC_MAX_WAIT_MS)
            .build()

        client.getCurrentLocation(request, CancellationTokenSource().token)
            .addOnSuccessListener { location ->
                val best = pickBestLocation(location, staleFallback)
                val payload = best?.let {
                    buildOneShotPayloadIfAccepted(
                        it,
                        includeBarometer = stopBarometerAfter,
                    )
                }
                if (payload != null) {
                    finishGetCurrentPosition(
                        stopBarometerAfter,
                        result,
                        payload,
                    )
                } else if (best != null) {
                    finishGetCurrentPosition(stopBarometerAfter, result, null)
                } else {
                    finishGetCurrentPosition(stopBarometerAfter, result, null)
                }
            }
            .addOnFailureListener { e ->
                if (staleFallback != null) {
                    val fallbackPayload = buildOneShotPayloadIfAccepted(
                        staleFallback,
                        includeBarometer = stopBarometerAfter,
                    )
                    finishGetCurrentPosition(
                        stopBarometerAfter,
                        result,
                        fallbackPayload,
                    )
                } else {
                    if (stopBarometerAfter && !running) {
                        barometer.stop()
                    }
                    result.error("GPS_ERROR", e.message ?: "getCurrentLocation failed", null)
                }
            }
    }

    private fun isLocationFresh(location: Location, maxAgeMs: Long): Boolean {
        val age = System.currentTimeMillis() - location.time
        return age in 0..maxAgeMs
    }

    private fun pickBestLocation(vararg locations: Location?): Location? {
        return locations.filterNotNull().minByOrNull { it.accuracy }
    }

    private fun buildOneShotPayloadIfAccepted(
        location: Location,
        includeBarometer: Boolean,
    ): Map<String, Any?>? {
        synchronized(this) {
            if (!fixGate.shouldAccept(location, SuperGpsFixQualityGate.FixSource.ONE_SHOT)) {
                return null
            }
            fixGate.noteAccepted(location)
        }
        return buildPayload(location, includeBarometer, applyKalman = true)
    }

    @Synchronized
    private fun start(context: Context) {
        if (running) return

        if (!hasFineLocation(context)) {
            emitError("PERMISSION_DENIED", "Location permission not granted")
            return
        }

        val client = LocationServices.getFusedLocationProviderClient(context)
        fusedClient = client
        kalman.reset()
        fixGate.reset()

        val appContext = context.applicationContext
        barometer.bind(appContext)
        val wantBaro = streamOptions.enableBarometer && barometer.hasHardwareSupport

        if (wantBaro) {
            barometer.reset()
            syncBarometer(appContext)
            val generation = ++startGeneration
            Thread {
                barometer.awaitFirstReading(SuperGpsBarometer.FIRST_READING_TIMEOUT_MS)
                Handler(Looper.getMainLooper()).post {
                    synchronized(this@SuperGpsLocationEngine) {
                        if (generation != startGeneration || sinks.isEmpty() || running) {
                            return@post
                        }
                        beginGpsStream(client)
                    }
                }
            }.start()
            return
        }

        syncBarometer(appContext)
        beginGpsStream(client)
    }

    @Synchronized
    private fun beginGpsStream(client: FusedLocationProviderClient) {
        if (running) return

        seedCachedLocations(client)

        val request = LocationRequest.Builder(
            Priority.PRIORITY_HIGH_ACCURACY,
            streamOptions.updateIntervalMs,
        )
            .setMinUpdateIntervalMillis(streamOptions.minUpdateIntervalMs)
            .setMinUpdateDistanceMeters(streamOptions.minUpdateDistanceMeters)
            .setMaxUpdateDelayMillis(0L)
            .setWaitForAccurateLocation(false)
            .build()

        val callback = object : LocationCallback() {
            override fun onLocationResult(result: LocationResult) {
                val location = result.lastLocation ?: return
                tryEmitLocation(location, SuperGpsFixQualityGate.FixSource.STREAM)
            }
        }
        locationCallback = callback

        try {
            client.requestLocationUpdates(
                request,
                callback,
                Looper.getMainLooper(),
            )
            running = true
        } catch (e: SecurityException) {
            emitError("PERMISSION_DENIED", e.message ?: "SecurityException")
            stopInternal()
        } catch (e: Exception) {
            emitError("GPS_ERROR", e.message ?: "Failed to start GPS")
            stopInternal()
        }
    }

    private fun seedCachedLocations(client: FusedLocationProviderClient) {
        client.lastLocation
            .addOnSuccessListener { location ->
                if (location != null) {
                    tryEmitLocation(location, SuperGpsFixQualityGate.FixSource.SEED_CACHE)
                }
            }

        val request = CurrentLocationRequest.Builder()
            .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
            .setMaxUpdateAgeMillis(30_000)
            .build()
        client.getCurrentLocation(request, CancellationTokenSource().token)
            .addOnSuccessListener { location ->
                if (location != null) {
                    tryEmitLocation(location, SuperGpsFixQualityGate.FixSource.SEED_CACHE)
                }
            }
    }

    @Synchronized
    private fun restartUpdates(context: Context) {
        stopInternal()
        kalman.reset()
        fixGate.reset()
        if (sinks.isNotEmpty()) {
            start(context)
        }
    }

    @Synchronized
    private fun stop() {
        stopInternal()
        kalman.reset()
        fixGate.reset()
    }

    private fun stopInternal() {
        startGeneration++
        val callback = locationCallback
        if (callback != null) {
            fusedClient?.removeLocationUpdates(callback)
        }
        locationCallback = null
        fusedClient = null
        running = false
        barometer.stop()
    }

    private fun syncBarometer(context: Context) {
        val appContext = context.applicationContext
        barometer.bind(appContext)
        if (streamOptions.enableBarometer && barometer.hasHardwareSupport) {
            barometer.start(appContext)
        } else {
            barometer.stop()
        }
    }

    private fun tryEmitLocation(location: Location, source: SuperGpsFixQualityGate.FixSource) {
        synchronized(this) {
            if (fixGate.shouldAccept(location, source)) {
                fixGate.noteAccepted(location)
                emitFilteredLocation(location)
                return
            }
            val fallback = fixGate.peekStreamFallbackLocation() ?: return
            fixGate.noteAccepted(fallback)
            emitFilteredLocation(fallback)
        }
    }

    private fun emitFilteredLocation(location: Location) {
        val includeBaro =
            streamOptions.enableBarometer && barometer.hasHardwareSupport
        val payload = buildPayload(location, includeBarometer = includeBaro)
        val snapshot = sinks.toList()
        for (sink in snapshot) {
            try {
                sink.success(payload)
            } catch (_: Exception) {
                // Sink có thể đã bị hủy.
            }
        }
    }

    private fun buildPayload(
        location: Location,
        includeBarometer: Boolean,
        applyKalman: Boolean = true,
    ): Map<String, Any?> {
        val (outLat, outLng) = if (applyKalman) {
            kalman.process(
                location.latitude,
                location.longitude,
                location.accuracy,
                location.speed.coerceAtLeast(0f),
                location.time,
            )
        } else {
            location.latitude to location.longitude
        }
        return locationToMap(location, outLat, outLng, includeBarometer)
    }

    private fun emitError(code: String, message: String) {
        val snapshot = sinks.toList()
        for (sink in snapshot) {
            try {
                sink.error(code, message, null)
            } catch (_: Exception) {
                // Bỏ qua.
            }
        }
    }

    private fun hasFineLocation(context: Context): Boolean {
        return ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION,
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun locationToMap(
        location: Location,
        filteredLat: Double,
        filteredLng: Double,
        includeBarometer: Boolean,
    ): Map<String, Any?> {
        val map = linkedMapOf<String, Any?>(
            "latitude" to filteredLat,
            "longitude" to filteredLng,
            "raw_latitude" to location.latitude,
            "raw_longitude" to location.longitude,
            "timestamp" to location.time,
            "accuracy" to location.accuracy.toDouble(),
            "altitude" to location.altitude,
            "heading" to location.bearing.toDouble(),
            "speed" to location.speed.toDouble(),
            "is_mocked" to isMocked(location),
        )

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            map["altitude_accuracy"] = location.verticalAccuracyMeters.toDouble()
            map["speed_accuracy"] = location.speedAccuracyMetersPerSecond.toDouble()
            map["heading_accuracy"] = location.bearingAccuracyDegrees.toDouble()
        } else {
            map["altitude_accuracy"] = 0.0
            map["speed_accuracy"] = 0.0
            map["heading_accuracy"] = 0.0
        }

        map["barometer_supported"] = barometer.hasHardwareSupport
        map["barometric_altitude"] = if (includeBarometer) {
            barometer.latestAltitudeM
        } else {
            null
        }

        return map
    }

    @Suppress("DEPRECATION")
    private fun isMocked(location: Location): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            location.isMock
        } else {
            location.isFromMockProvider
        }
    }
}

File: ios/Runner/SuperGpsLocationEngine.swift
import CoreLocation
import Flutter
import Foundation

/// CoreLocation + Kalman — payload tương thích Android SuperGpsLocationEngine.
final class SuperGpsLocationEngine: NSObject, CLLocationManagerDelegate {
    static let shared = SuperGpsLocationEngine()

    private static let fastCacheMaxAgeMs: Int64 = 120_000
    private static let currentLocMaxAgeMs: Int64 = 60_000
    private static let currentLocMaxWaitMs: Int64 = 1_500
    private static let oneShotBestAccuracyM = 4.0

    private let kalman = SuperGpsKalmanFilter()
    private let fixGate = SuperGpsFixQualityGate()
    private let barometer = SuperGpsBarometer()
    private let lock = NSLock()

    private var locationManager: CLLocationManager?
    private var sinks: [FlutterEventSink] = []
    private var streamOptions = SuperGpsStreamOptions()
    private var running = false
    private var startGeneration = 0

    private var pendingOneShotResult: FlutterResult?
    private var pendingStopBarometerAfterOneShot = false
    private var oneShotTimeoutWorkItem: DispatchWorkItem?

    private override init() {
        super.init()
    }

    func isBarometerHardwareSupported() -> Bool {
        barometer.hasHardwareSupport
    }

    func addListener(sink: @escaping FlutterEventSink, options: SuperGpsStreamOptions) {
        lock.lock()
        defer { lock.unlock() }

        let optionsChanged = streamOptions != options
        streamOptions = options
        sinks.append(sink)

        if !running {
            startUpdates()
        } else if optionsChanged {
            restartUpdates()
        }
    }

    func removeAllListeners() {
        lock.lock()
        defer { lock.unlock() }

        sinks.removeAll()
        if sinks.isEmpty {
            stopUpdates()
        }
    }

    func getCurrentPosition(enableBarometer: Bool, result: @escaping FlutterResult) {
        guard hasLocationPermission() else {
            result(
                FlutterError(
                    code: "PERMISSION_DENIED",
                    message: "Location permission not granted",
                    details: nil
                )
            )
            return
        }

        let shouldRunBaro = enableBarometer && barometer.hasHardwareSupport
        let manager = ensureLocationManager()

        if shouldRunBaro {
            if !barometer.isActive {
                barometer.start()
            }
            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                guard let self else { return }
                _ = self.barometer.awaitFirstReading(
                    timeoutMs: SuperGpsBarometer.firstReadingTimeoutMs
                )
                DispatchQueue.main.async {
                    self.fetchCurrentPositionGps(
                        manager: manager,
                        shouldRunBaro: true,
                        result: result
                    )
                }
            }
            return
        }

        fetchCurrentPositionGps(
            manager: manager,
            shouldRunBaro: false,
            result: result
        )
    }

    private func fetchCurrentPositionGps(
        manager: CLLocationManager,
        shouldRunBaro: Bool,
        result: @escaping FlutterResult
    ) {
        lock.lock()
        if !running {
            kalman.reset()
            fixGate.reset()
        }
        lock.unlock()

        if let cached = manager.location,
           isLocationFresh(cached, maxAgeMs: Self.fastCacheMaxAgeMs),
           horizontalAccuracyM(cached) <= Self.oneShotBestAccuracyM,
           let payload = buildOneShotPayloadIfAccepted(
               cached,
               includeBarometer: shouldRunBaro
           ) {
            finishGetCurrentPosition(
                stopBarometerAfter: shouldRunBaro,
                result: result,
                payload: payload
            )
            return
        }

        pendingOneShotResult = result
        pendingStopBarometerAfterOneShot = shouldRunBaro

        oneShotTimeoutWorkItem?.cancel()
        let timeoutItem = DispatchWorkItem { [weak self] in
            self?.completeOneShotWithFallback(manager: manager, enableBarometer: shouldRunBaro)
        }
        oneShotTimeoutWorkItem = timeoutItem
        DispatchQueue.main.asyncAfter(
            deadline: .now() + Double(Self.currentLocMaxWaitMs) / 1000.0,
            execute: timeoutItem
        )

        manager.requestLocation()
    }

    // MARK: - CLLocationManagerDelegate

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }

        if let pending = takePendingOneShotResult() {
            let best = pickBestLocation(location, manager.location) ?? location
            let payload = buildOneShotPayloadIfAccepted(
                best,
                includeBarometer: pendingStopBarometerAfterOneShot
            )
            finishGetCurrentPosition(
                stopBarometerAfter: pendingStopBarometerAfterOneShot,
                result: pending,
                payload: payload
            )
            return
        }

        tryEmitLocation(location, source: .stream)
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        if let pending = takePendingOneShotResult() {
            if let cached = manager.location {
                let payload = buildOneShotPayloadIfAccepted(
                    cached,
                    includeBarometer: pendingStopBarometerAfterOneShot
                )
                finishGetCurrentPosition(
                    stopBarometerAfter: pendingStopBarometerAfterOneShot,
                    result: pending,
                    payload: payload
                )
            } else {
                if pendingStopBarometerAfterOneShot && !running {
                    barometer.stop()
                }
                pending(
                    FlutterError(
                        code: "GPS_ERROR",
                        message: error.localizedDescription,
                        details: nil
                    )
                )
            }
            return
        }

        emitError(code: "GPS_ERROR", message: error.localizedDescription)
    }

    // MARK: - Private

    private func completeOneShotWithFallback(manager: CLLocationManager, enableBarometer: Bool) {
        guard let pending = takePendingOneShotResult() else { return }

        if let cached = manager.location {
            let payload = buildOneShotPayloadIfAccepted(
                cached,
                includeBarometer: enableBarometer
            )
            finishGetCurrentPosition(
                stopBarometerAfter: enableBarometer,
                result: pending,
                payload: payload
            )
        } else {
            finishGetCurrentPosition(
                stopBarometerAfter: enableBarometer,
                result: pending,
                payload: nil
            )
        }
    }

    private func takePendingOneShotResult() -> FlutterResult? {
        oneShotTimeoutWorkItem?.cancel()
        oneShotTimeoutWorkItem = nil
        let pending = pendingOneShotResult
        pendingOneShotResult = nil
        return pending
    }

    private func finishGetCurrentPosition(
        stopBarometerAfter: Bool,
        result: @escaping FlutterResult,
        payload: [String: Any?]?
    ) {
        if stopBarometerAfter && !running {
            barometer.stop()
        }
        result(payload)
    }

    private func startUpdates() {
        guard hasLocationPermission() else {
            emitError(code: "PERMISSION_DENIED", message: "Location permission not granted")
            return
        }

        let manager = ensureLocationManager()
        kalman.reset()
        fixGate.reset()
        applyStreamOptions(to: manager)

        let wantBaro = streamOptions.enableBarometer && barometer.hasHardwareSupport
        lock.lock()
        startGeneration += 1
        let generation = startGeneration
        lock.unlock()

        if wantBaro {
            barometer.reset()
            syncBarometer(with: manager.location)
            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                guard let self else { return }
                _ = self.barometer.awaitFirstReading(
                    timeoutMs: SuperGpsBarometer.firstReadingTimeoutMs
                )
                DispatchQueue.main.async {
                    self.lock.lock()
                    defer { self.lock.unlock() }
                    guard self.startGeneration == generation,
                          !self.sinks.isEmpty,
                          !self.running else {
                        return
                    }
                    self.beginGpsStream(manager)
                }
            }
            return
        }

        syncBarometer(with: manager.location)
        beginGpsStream(manager)
    }

    private func beginGpsStream(_ manager: CLLocationManager) {
        guard !running else { return }

        if let cached = manager.location {
            tryEmitLocation(cached, source: .seedCache)
        }

        manager.startUpdatingLocation()
        running = true
    }

    private func restartUpdates() {
        stopInternal()
        kalman.reset()
        fixGate.reset()
        if !sinks.isEmpty {
            startUpdates()
        }
    }

    private func stopUpdates() {
        stopInternal()
        kalman.reset()
        fixGate.reset()
    }

    private func stopInternal() {
        lock.lock()
        startGeneration += 1
        lock.unlock()

        locationManager?.stopUpdatingLocation()
        running = false
        barometer.stop()
    }

    private func ensureLocationManager() -> CLLocationManager {
        if let manager = locationManager { return manager }

        let manager = CLLocationManager()
        manager.delegate = self
        manager.desiredAccuracy = kCLLocationAccuracyBestForNavigation
        manager.activityType = .fitness
        manager.pausesLocationUpdatesAutomatically = false
        locationManager = manager
        return manager
    }

    private func applyStreamOptions(to manager: CLLocationManager) {
        manager.distanceFilter = CLLocationDistance(streamOptions.minUpdateDistanceMeters)
    }

    private func syncBarometer(with location: CLLocation?) {
        if streamOptions.enableBarometer && barometer.hasHardwareSupport {
            if let location, location.altitude.isFinite {
                barometer.setGpsBaselineMeters(location.altitude)
            }
            barometer.start()
        } else {
            barometer.stop()
        }
    }

    private func tryEmitLocation(_ location: CLLocation, source: SuperGpsFixQualityGate.FixSource) {
        lock.lock()
        if fixGate.shouldAccept(location, source: source) {
            fixGate.noteAccepted(location)
            lock.unlock()
            emitFilteredLocation(location)
            return
        }
        if let fallback = fixGate.peekStreamFallbackLocation() {
            fixGate.noteAccepted(fallback)
            lock.unlock()
            emitFilteredLocation(fallback)
            return
        }
        lock.unlock()
    }

    private func emitFilteredLocation(_ location: CLLocation) {
        if streamOptions.enableBarometer && barometer.hasHardwareSupport {
            barometer.setGpsBaselineMeters(
                location.altitude.isFinite ? location.altitude : nil
            )
        }

        let includeBaro = streamOptions.enableBarometer && barometer.hasHardwareSupport
        let payload = buildPayload(location, includeBarometer: includeBaro)

        lock.lock()
        let snapshot = sinks
        lock.unlock()

        for sink in snapshot {
            sink(payload)
        }
    }

    private func buildOneShotPayloadIfAccepted(
        _ location: CLLocation,
        includeBarometer: Bool
    ) -> [String: Any?]? {
        lock.lock()
        let accept = fixGate.shouldAccept(location, source: .oneShot)
        if accept {
            fixGate.noteAccepted(location)
        }
        lock.unlock()
        guard accept else { return nil }
        return buildPayload(location, includeBarometer: includeBarometer, applyKalman: true)
    }

    private func pickBestLocation(_ locations: CLLocation...) -> CLLocation? {
        locations.min { horizontalAccuracyM($0) < horizontalAccuracyM($1) }
    }

    private func horizontalAccuracyM(_ location: CLLocation) -> Double {
        let accuracy = location.horizontalAccuracy
        return accuracy >= 0 ? accuracy : Double.greatestFiniteMagnitude
    }

    private func buildPayload(
        _ location: CLLocation,
        includeBarometer: Bool,
        applyKalman: Bool = true
    ) -> [String: Any?] {
        let accuracy = max(location.horizontalAccuracy, 0)
        let coordinate: (Double, Double)
        if applyKalman {
            coordinate = kalman.process(
                latitude: location.coordinate.latitude,
                longitude: location.coordinate.longitude,
                accuracyM: accuracy,
                speedMps: max(location.speed, 0),
                timestampMs: Int64(location.timestamp.timeIntervalSince1970 * 1000)
            )
        } else {
            coordinate = (
                location.coordinate.latitude,
                location.coordinate.longitude
            )
        }

        var map: [String: Any?] = [
            "latitude": coordinate.0,
            "longitude": coordinate.1,
            "raw_latitude": location.coordinate.latitude,
            "raw_longitude": location.coordinate.longitude,
            "timestamp": Int64(location.timestamp.timeIntervalSince1970 * 1000),
            "accuracy": accuracy,
            "altitude": location.altitude,
            "heading": location.course >= 0 ? location.course : 0.0,
            "speed": max(location.speed, 0),
            "is_mocked": isMocked(location),
            "altitude_accuracy": location.verticalAccuracy >= 0
                ? location.verticalAccuracy
                : 0.0,
            "speed_accuracy": location.speedAccuracy >= 0
                ? location.speedAccuracy
                : 0.0,
            "heading_accuracy": location.courseAccuracy >= 0
                ? location.courseAccuracy
                : 0.0,
            "barometer_supported": barometer.hasHardwareSupport,
            "barometric_altitude": includeBarometer ? barometer.latestAltitudeM : nil,
        ]

        return map
    }

    private func emitError(code: String, message: String) {
        lock.lock()
        let snapshot = sinks
        lock.unlock()

        for sink in snapshot {
            sink(FlutterError(code: code, message: message, details: nil))
        }
    }

    private func hasLocationPermission() -> Bool {
        let status = ensureLocationManager().authorizationStatus
        return status == .authorizedWhenInUse || status == .authorizedAlways
    }

    private func isLocationFresh(_ location: CLLocation, maxAgeMs: Int64) -> Bool {
        let ageMs = Int64(Date().timeIntervalSince(location.timestamp) * 1000)
        return ageMs >= 0 && ageMs <= maxAgeMs
    }

    private func isMocked(_ location: CLLocation) -> Bool {
        if #available(iOS 15.0, *) {
            return location.sourceInformation?.isSimulatedBySoftware == true
        }
        return false
    }
}
